src/mame/marion/ms68k.cpp

```diff
@@ -0,0 +1,319 @@
+			// SCSI Interrupt = 5
+			downcast<ncr53c80_device &>(*device).irq_handler().set_inputline(m_maincpu, M68K_IRQ_5);
+		});
+
+	// Set up FDC.
+
+	WD1772(config, m_fdc, 16_MHz_XTAL / 2);
+	// FDC Interrupt = 4
+	m_fdc->intrq_wr_callback().set_inputline(m_maincpu, M68K_IRQ_3);
+	// DUART OP3 = Side 1 Select
+	m_duart->outport_cb().append(FUNC(ms68k_state::floppy_side_w)).bit(3);
+	// DUART OP4 = Drive Select 0
+	m_duart->outport_cb().append(FUNC(ms68k_state::floppy_drive0_w)).bit(4);
+	// DUART OP5 = Drive Select 1
+	m_duart->outport_cb().append(FUNC(ms68k_state::floppy_drive1_w)).bit(5);
+	// DUART OP6 = Double Density
+	m_duart->outport_cb().append(m_fdc, FUNC(wd1772_device::dden_w)).bit(6);
+	FLOPPY_CONNECTOR(config, m_floppy[0], floppy_types, "525sd", floppy_formats);
+	FLOPPY_CONNECTOR(config, m_floppy[1], floppy_types, "525sd", floppy_formats);
+
+	// Set up printer.
+
+	CENTRONICS(config, m_printer_conn, centronics_devices, "printer");
+	OUTPUT_LATCH(config, m_printer_out);
+	m_printer_conn->set_output_latch(*m_printer_out);
+	// Printer Interrupt = 2
+	m_printer_conn->ack_handler().append_inputline(m_maincpu, M68K_IRQ_2);
+	// DUART IP2 = Busy
+	m_printer_conn->busy_handler().set(m_duart, FUNC(mc68681_device::ip2_w));
+	// DUART IP3 = Paper Out
+	m_printer_conn->perror_handler().set(m_duart, FUNC(mc68681_device::ip3_w));
+	// DUART IP4 = Device Selected
+	m_printer_conn->select_handler().set(m_duart, FUNC(mc68681_device::ip4_w));
+	// DUART IP5 = Acknowledge
+	m_printer_conn->ack_handler().set(m_duart, FUNC(mc68681_device::ip5_w));
+
+	// DUART OP2 = Initialize
+	m_duart->outport_cb().append(m_printer_conn, FUNC(centronics_device::write_init)).bit(2);
+	// DUART OP7 = Data Strobe
+	m_duart->outport_cb().append(m_printer_conn, FUNC(centronics_device::write_strobe)).bit(7);
+
+	// Set up Expansion.
+	// EXT Interrupt = 6
+	// TODO: Implement expansion bus.
+}
+
+void ms68k_state::mem_map(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x000000, 0x07ffff).ram().share("ram"); // 512KB RAM onboard maximum
+	map(0x000008, 0x07ffff).ram();
+	map(0xd00000, 0xd7ffff).rw(m_scsi, FUNC(ncr5380_device::read), FUNC(ncr5380_device::write)).umask16(0x00ff);
+	map(0xd80000, 0xdfffff).rw(FUNC(ms68k_state::printer_r), FUNC(ms68k_state::printer_w)).umask16(0x00ff);
+	map(0xe00000, 0xe7ffff).rw(m_fdc, FUNC(wd1772_device::read), FUNC(wd1772_device::write)).umask16(0x00ff);
+	map(0xe80000, 0xefffff).rw(FUNC(ms68k_state::duart_r), FUNC(ms68k_state::duart_w)).umask16(0x00ff);
+	map(0xf00000, 0xf1ffff).rom().region("roms", 0); // 128KB ROM onboard maximum
+
+	// ROM is mapped to 0 on reset, remapped on first DUART write
+	map(0x000000, 0x000007).view(m_bootvect);
+	m_bootvect[0](0x000000, 0x000007).rom().region("roms", 0);
+}
+
+void ms68k_state::machine_reset()
+{
+	// Reset pointer to bootvector in ROM for bootvector view
+	m_bootvect.select(0);
+}
+
+/* DUART */
+
+uint8_t ms68k_state::duart_r(offs_t offset)
+{
+	return m_duart->read(offset);
+}
```

Comment:
You shouldn’t need this trampoline – you should be able to do something like:
```c++
	map(0xe80000, 0xefffff).r(m_duart, FUNC(mc68681_device::read)).w(FUNC(ms68k_state::duart_w)).umask16(0x00ff);
```

Reply:
Oh, sick, that’s a lot more convenient! Thanks!